flexmeasures/auth/decorators.py

```diff
@@ -102,3 +98,82 @@ def decorated_view(*args, **kwargs):
         return decorated_view
 
     return wrapper
+
+
+def permission_required_for_context(
+    permission: str, arg_pos: Optional[int] = None, arg_name: Optional[str] = None
+):
+    """
+    This decorator can be used to make sure that the current user has the necessary permission to access the context.
+    The context needs to be an AuthModelMixin and is found in the keyword arguments by name and/or by a position in the non-keyword arguments (defaults to 0).
+    Using both arguments is useful when Marshmallow places a dict of de-serialized fields and you are using use_args.
+    The permission needs to be a known permission and is checked with principal descriptions from the context's access control list (see AuthModelMixin.__acl__).
+
+    Usually, you'd place a marshmallow field further up in the decorator chain, e.g.:
+
+        @app.route("/resource/<resource_id>", methods=["GET"])
+        @use_kwargs(
+            {"the_resource": ResourceIdField(data_key="resource_id")},
+            location="path",
+        )
+        @permission_required_for_context("read", arg_name="the_resource")
+        @as_json
+        def view(resource_id: int, resource: Resource):
+            return dict(name=resource.name)
+
+    Where `ResourceIdField._deserialize()` turns the id parameter into a resource context (if possible).
+
+    This decorator raises a 403 response if there is no principal for the required permission.
+    It raises a 401 response if the user is not authenticated at all.
```

Comment:
Where is the decorator returning a User? This decorator also doesn't alter the arguments.

I believe by returning `_security._unauthn_handler()` I am following the way Flask-Security also implements their decorators:

- https://flask-security-too.readthedocs.io/en/stable/customizing.html#oh-my
- https://flask-security-too.readthedocs.io/en/stable/api.html#flask_security.Security.unauthz_handler

Anyways, as `flexmeasuresauth.__init__` shows, raising 403 reults directly in this unauthz_handler, so in our own code we can also treat this as a matter of style.

Besides the advantage that we pass on which decorator initiates the 403 problem. Our permission decorator is logging a  distinct error message before each call to the unauthz_handler, but I'm thinking if we should rather log errors in the central error handler.

Reply:
> Where is the decorator returning a User? This decorator also doesn't alter the arguments.

Indeed. I was wrong here. The decorator gives back the called function (that was decorated) or a response tuple.

> Anyways, as flexmeasures.auth.__init__ shows, raising 403 results directly in this unauthz_handler, so in our own code we can also treat this as a matter of style.

as I understand it, in case the decorated function returns a response tuple rather than raise, the decorator should only be used on functions that are themselves expected to return a response tuple. It seems to me that this mean the `permission_required_for_context` can not be used to decorate e.g. routes in the UI package that return an html.

On the other hand, if the decorator would raise an exception that should mean `unauthenticated_handler_html`, being the registered handler for the UI package, would pick up the exception.

> Besides the advantage that we pass on which decorator initiates the 403 problem.

I don't see the difference between adding that info to a response tuple or to an exception. In the latter case, I would expect the exception handler to copy the error message to the response.

> I'm thinking if we should rather log errors in the central error handler.

What would the code look like to "log errors in the central error handler"?

Reply:
>  in case the decorated function returns a response tuple rather than raise, the decorator should only be used on functions that are themselves expected to return a response tuple.
> if the decorator would raise an exception that should mean `unauthenticated_handler_html` ... 

If any code anywhere in FM raises a 403 error, our registered `unauthorized_handler` comes in (because our `unauthorized_handler_e` directly calls it). But it *only* returns a response tuple `if request.is_json or request.content_type is None`. UI requests need to say they want HTML back.
This policy preference is a bit new. I'm not saying it's perfect right now, but I see it as an improvement. It implements our strategy of regarding the FM API a first-class citizen.

> What would the code look like to "log errors in the central error handler"?
One call to `current_app.logger.error` in the `unauthorized_handler_e` instead of many in the endpoints.
I'm considering at this point that it might be a good policy that our own code raises 403 errors with a good message. Our error handler logs the message, but we do not display it to the user (usually auth systems don't want to give out information how they work). flask_security's decorators are calling the unath_handler directly as they don't know if an auth error handler is registered/ Maybe they otherwise would also rather raise, I don't know.